### PR TITLE
ECL-173: ECL initial declarative registration journey tests

### PR DESCRIPTION
**Pull request by jamesvarghese701**

### CODE REVIEW

File: run_registration.sh [hunk 1]
```diff
@@ -1,18 +1,18 @@
...
     DRIVER="-Dwebdriver.gecko.driver=/usr/local/bin/geckodriver"
 fi
 
 
 # Scalafmt checks have been separated from the test command to avoid OutOfMemoryError in Jenkins
-sbt scalafmtCheckAll scalafmtSbtCheck
+# sbt scalafmtCheckAll scalafmtSbtCheck
 
-sbt -Dbrowser="${BROWSER_TYPE:=$BROWSER}" -Denvironment="${ENV:=local}" "testOnly uk.gov.hmrc.test.ui.cucumber.runner.RunRegistration"
+sbt -Dbrowser=$BROWSER -Denvironment=$ENV $DRIVER "testOnly uk.gov.hmrc.test.ui.cucumber.runner.RunRegistration"
```
Comment (peter-hazell) [inline run_registration.sh:18]: Can we try and keep the defaults of `chrome` and `local` please? This is handy when wanting to run the tests quickly, can't see a reason why it shouldn't work?

File: src/test/resources/application.conf [hunk 2]
```diff
@@ -17,9 +17,13 @@ env: local
...
   services {
     host: "http://localhost"
-    auth-login-stub {
-      port: 9949
-      productionRoute = "/auth-login-stub"
+    auth-login-stub-reg {
+      port: 14000
+      productionRoute = ""
+    }
+    auth-login-stub-ret {
```
Comment (peter-hazell) [inline src/test/resources/application.conf:24]: Let's rename these - they are no longer the auth-login-stub .. they are the start pages for the relevant services

so they should be:

`economic-crime-levy-registration-frontend`
`economic-crime-levy-returns-frontend`

File: src/test/resources/features/Registration.feature [hunk 3]
```diff
@@ -0,0 +1,14 @@
+@all @reg @accessibility @zap
+Feature: Register for ECL service
+
+  Scenario Outline: User able to select their UK revenue category for ECL service registration
+    Given I am on the registration start page
+    When I click on Start now button
+    And I enter valid login credentials
```
Comment (peter-hazell) [inline src/test/resources/features/Registration.feature:7]: As discussed - these test packs should initially be automated against service manager / local, not QA

we can consider some automated QA smoke tests in the future once it is in a bit more of a stable/known state - but this relies on real data.

File: src/test/scala/uk/gov/hmrc/test/ui/cucumber/stepdefs/RegistrationStepDef.scala [hunk 4]
```diff
@@ -17,19 +17,33 @@
...
 
-  When("""I click on submit button""") { () =>
+  When("""^I click on (.*) button""") { (value: String) =>
+    SharedActions.clickLinkByPartialText(value)
+  }
+
+  And("""^I enter valid login credentials""") { () =>
+    GatewaySignInPage.signInGateway(userID = "67 03 89 79 83 09", password = "Opencast123")
+    GatewaySignInPage.enterAccessCode(accessCode = "123456")
+  }
```
Comment (peter-hazell) [inline src/test/scala/uk/gov/hmrc/test/ui/cucumber/stepdefs/RegistrationStepDef.scala:36]: Let's remove this for now - until we include QA tests in the future

File: src/test/scala/uk/gov/hmrc/test/ui/pages/SharedActions.scala [hunk 5]
```diff
@@ -17,14 +17,41 @@
...
+      assertPage("Select your UK turnover for {0}")
+  }
+    else {
+    click(By.xpath(s"//label[contains(text(), '$value')]"))
+    driver.findElement(By.cssSelector(".govuk-button")).click()
+    value match {
+        case "Less than £10.2 million" => assertPage("You do not need to register for the Economic Crime Levy")
+        case "Equal to or more than £10.2 million" => assertPage("Who is your Anti-Money Laundering (AML) supervisor?")
+    }
+    }
```
Comment (peter-hazell) [inline src/test/scala/uk/gov/hmrc/test/ui/pages/SharedActions.scala:54]: Not quite sure why you need this 

The radio buttons have IDs assigned to them:

`less-than` and `equal-or-more-than`, so can you not use these to select ?

File: src/test/scala/uk/gov/hmrc/test/ui/pages/registration/GatewaySignInPage.scala [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+    sendKeys(By.id("user_id"), userID)
+    sendKeys(By.id("password"), password)
+    click(By.id("continue"))
+
+  }
+
+  def enterAccessCode(accessCode: String): Unit = {
+    sendKeys(By.xpath("//input[@id='oneTimePassword']"), accessCode)
+    click(By.id("continue"))
+  }
```
Comment (peter-hazell) [inline src/test/scala/uk/gov/hmrc/test/ui/pages/registration/GatewaySignInPage.scala:42]: Don't think we need this for now as discussed

File: src/test/resources/features/Registration.feature [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+    And I select a <UK revenue> category for ECL payments
```
Comment (peter-hazell) [inline src/test/resources/features/Registration.feature:8]:
```suggestion
    And I select that my UK revenue is <UK revenue>
```

File: src/test/resources/features/Registration.feature [hunk 8]
```diff
@@ -0,0 +1,14 @@
+@all @reg @accessibility @zap
+Feature: Register for ECL service
+
+  Scenario Outline: User able to select their UK revenue category for ECL service registration
```
Comment (peter-hazell) [inline src/test/resources/features/Registration.feature:4]:
```suggestion
  Scenario Outline: User able to select whether their UK revenue is less than or more than/equal to £10.2m
```

File: run_registration.sh (hunk 1 above)
Comment (jamesvarghese701) [reply run_registration.sh]: changed.

File: src/test/resources/application.conf (hunk 2 above)
Comment (jamesvarghese701) [reply src/test/resources/application.conf]: done

File: src/test/resources/features/Registration.feature (hunk 3 above)
Comment (jamesvarghese701) [reply src/test/resources/features/Registration.feature]: changed the code to run against local environment

File: src/test/scala/uk/gov/hmrc/test/ui/cucumber/stepdefs/RegistrationStepDef.scala (hunk 4 above)
Comment (jamesvarghese701) [reply src/test/scala/uk/gov/hmrc/test/ui/cucumber/stepdefs/RegistrationStepDef.scala]: removed

File: src/test/scala/uk/gov/hmrc/test/ui/pages/SharedActions.scala (hunk 5 above)
Comment (jamesvarghese701) [reply src/test/scala/uk/gov/hmrc/test/ui/pages/SharedActions.scala]: I have used if else logic, any suggestion?

File: src/test/scala/uk/gov/hmrc/test/ui/pages/SharedActions.scala [hunk 9]
```diff
@@ -17,14 +17,41 @@
 package uk.gov.hmrc.test.ui.pages
 
 import org.openqa.selenium.By
-import uk.gov.hmrc.test.ui.pages.registration.RegistrationLoginPage.authoritySubmitButton
 
 object SharedActions extends BasePage {
 
+  val authoritySubmitButton = By.id("submit-top")
```
[the remark this answers is not in the document]
Comment (jamesvarghese701) [reply src/test/scala/uk/gov/hmrc/test/ui/pages/SharedActions.scala]: refactored the code.

File: src/test/scala/uk/gov/hmrc/test/ui/pages/registration/GatewaySignInPage.scala (hunk 6 above)
Comment (jamesvarghese701) [reply src/test/scala/uk/gov/hmrc/test/ui/pages/registration/GatewaySignInPage.scala]: removed

File: src/test/resources/features/Registration.feature (hunk 7 above)
Comment (jamesvarghese701) [reply src/test/resources/features/Registration.feature]: changed

File: src/test/resources/features/Registration.feature (hunk 8 above)
Comment (jamesvarghese701) [reply src/test/resources/features/Registration.feature]: changed

File: src/test/resources/application.conf [hunk 10]
```diff
@@ -17,9 +17,13 @@ env: local
...
   services {
     host: "http://localhost"
-    auth-login-stub {
-      port: 9949
-      productionRoute = "/auth-login-stub"
+    economic-crime-levy-registration-frontend {
+      port: 14000
+      productionRoute = ""
+    }
+    auth-login-stub-ret {
```
Comment (peter-hazell) [inline src/test/resources/application.conf:24]: same as above - this should be the name of the service

`economic-crime-levy-returns-frontend`

File: src/test/scala/uk/gov/hmrc/test/ui/pages/SharedActions.scala [hunk 11]
```diff
@@ -17,14 +17,34 @@
...
+    else if (value == "less-than"){
+      clickById(value)
+      findElementByCssSelector(".govuk-button").click()
+      assertPartialTextIsDisplayed("You do not need to register for the Economic Crime Levy")
+    }
+    else{
+      clickById(value)
+      findElementByCssSelector(".govuk-button").click()
+      assertPartialTextIsDisplayed("Who is your Anti-Money Laundering (AML) supervisor?")
+    }
```
Comment (peter-hazell) [inline src/test/scala/uk/gov/hmrc/test/ui/pages/SharedActions.scala:48]: this is doing assertions as well as clicking buttons

1. this is a shared action so shouldn't contain lots of branching logic like this.. you should use the shared action from wherever it relates to, that's the reason it's a shared action
2. should this be doing assertions? the assertion should be being made from wherever it's being called

Comment (peter-hazell) [reply src/test/scala/uk/gov/hmrc/test/ui/pages/SharedActions.scala]: Addressed via pairing

File: src/test/resources/features/Registration.feature [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+      | otherProfessionalBody__option--1  | Limited liability partnership | Stub GRS Journey Data                            |
+      | otherProfessionalBody__option--5  | Limited partnership           | Stub GRS Journey Data                            |
+      | otherProfessionalBody__option--10 | Scottish limited partnership  | Stub GRS Journey Data                            |
+      | otherProfessionalBody__option--15 | General partnership           | Stub GRS Journey Data                            |
+      | otherProfessionalBody__option--20 | Scottish partnership          | Stub GRS Journey Data                            |
+      | otherProfessionalBody__option--19 | Sole trader                   | Stub GRS Journey Data                            |
+      | otherProfessionalBody__option--18 | Other                         | Sorry, we’re experiencing technical difficulties |
+
+
+  Scenario Outline: User is able to select their AML supervisor as HMRC and proceed to the next page
```
Comment (peter-hazell) [inline src/test/resources/features/Registration.feature:71]: The scenario description seems wrong - this should be about selecting entity type, not AML supervisor?

File: src/test/resources/features/Registration.feature [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+    Given I am on the AML supervisor page
+    And I click on the Submit button on the authority wizard page
```
Comment (peter-hazell) [inline src/test/resources/features/Registration.feature:73]: This doesn't make sense, as the 'Given' says you are on the AML supervisor page, so the sign in should be before this?

File: src/test/resources/features/Registration.feature (hunk 12 above)
Comment (jamesvarghese701) [reply src/test/resources/features/Registration.feature]: Done

File: src/test/resources/features/Registration.feature [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+    And I click on the Save and continue button
+    Then I should be on the page with the content <Expected content>
+
+    Examples:
+      | AML supervisor              | Expected content                                                                |
+      | HMRC                        | What is your entity type?                                                       |
+      | Gambling Commission         | You need to register with the Gambling Commission (GC) to pay the levy          |
+      | Financial Conduct Authority | You need to register with the Financial Conduct Authority (FCA) to pay the levy |
+
+  Scenario Outline: User is able to select their AML supervisor as Other and proceed to the next page
```
Comment (peter-hazell) [inline src/test/resources/features/Registration.feature:49]: this description doesn't align with the test that is being done, this says AML supervisor but it's entity type selection

**Review by peter-hazell — APPROVED** [review]

**Pull request merged**